pkg/csi/hostpathimpl.go

```diff
@@ -0,0 +1,264 @@
+/*
+Copyright (c) 2023 ApeCloud, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	log "k8s.io/klog/v2"
+
+	"github.com/alibaba/open-local/pkg"
+)
+
+const (
+	hostPathBasePathEnv = "HOSTPATH_BASE_PATH"
+
+	// tags from StorageClass
+	hostPathScTag             = "hostPath"
+	enforceCapacityLimitScTag = "enforceCapacityLimit"
+
+	// internal tags, the prefix is added to avoid conflicts with other public tags
+	volumeCapacityTag = "_hp_volumeCapacity"
+)
+
+// hostPathImpl implements the CSI interfaces for hostPath volume type
+type hostPathImpl struct {
+	baseControllerServer
+	baseNodeServer
+
+	// TODO(x.zhou): needs further refactoring to move methods/fields
+	//               from controllerServer/nodeServer to baseControllerServer/baseNodeServer.
+	cs *controllerServer
+	ns *nodeServer
+}
+
+func (impl *hostPathImpl) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	if volumeSource := req.GetVolumeContentSource(); volumeSource != nil {
+		return nil, status.Error(codes.Unimplemented, "CreateVolume: volume content source is not supported")
+	}
+
+	reqCtx := getValue(ctx, ctxKeyCreateVolume)
+	nodeName := reqCtx.nodeName
+	volumeID := req.GetName()
+	parameters := req.GetParameters()
+
+	// TODO(x.zhou): interact with the scheduler to see if there is sufficient space to allocate
```

Comment:
if there is no sufficient space to allocate this volume, we can only return error, then the user will be stucked here, so we need extend schedule intead of check here?

Am I right?